### PR TITLE
Feat/lp token

### PR DESCRIPTION
**Pull request by senichevalexey**

### CODE REVIEW

**Comment by senichevalexey** [comment]
It was Mulili`s PR, I just recovered it

File: contracts/ZunamiVault.sol [hunk 1]
```diff
@@ -1,13 +1,101 @@
...
+    }
+
+    // Custom logic in here for how much the vault allows to be borrowed
+    // Sets minimum required on-hand to keep small withdrawals cheap
+    function available() public view returns (uint256) {
+        return token.balanceOf(address(this));
+    }
+
+    function depositAll() external {
+        deposit(token.balanceOf(msg.sender));
```
Comment (smartdev0725) [inline contracts/ZunamiVault.sol:57]: I think this function needs to have a require to check msg.sender's balance.
For example: require(token.balanceOf(msg.sender) > 0, "Insufficient amount for deposit");

File: contracts/ZunamiVault.sol [hunk 2]
```diff
@@ -1,13 +1,101 @@
...
+        if (totalSupply() == 0) {
+            shares = _amount;
+        } else {
+            shares = (_amount.mul(totalSupply())).div(_pool);
+        }
+        _mint(msg.sender, shares);
+    }
+
+    function withdrawAll() external {
+        withdraw(balanceOf(msg.sender));
```
Comment (smartdev0725) [inline contracts/ZunamiVault.sol:74]: require(balanceOf(msg.sender) > 0, "Insufficient amount");

File: contracts/ZunamiVault.sol [hunk 3]
```diff
@@ -1,13 +1,101 @@
...
+        uint256 shares = 0;
+        if (totalSupply() == 0) {
+            shares = _amount;
+        } else {
+            shares = (_amount.mul(totalSupply())).div(_pool);
+        }
+        _mint(msg.sender, shares);
+    }
+
+    function withdrawAll() external {
```
Comment (smartdev0725) [inline contracts/ZunamiVault.sol:73]: require(balanceOf(msg.sender) > 0, "Insufficient amount");

File: contracts/ZunamiVault.sol [hunk 4]
```diff
@@ -1,13 +1,101 @@
...
+        return token.balanceOf(address(this)).add(IController(controller).balanceOf(address(token)));
+    }
+
+    // Custom logic in here for how much the vault allows to be borrowed
+    // Sets minimum required on-hand to keep small withdrawals cheap
+    function available() public view returns (uint256) {
+        return token.balanceOf(address(this));
+    }
+
+    function depositAll() external {
```
Comment (smartdev0725) [inline contracts/ZunamiVault.sol:56]: I think this function needs to have a require to check msg.sender's balance.
For example: require(token.balanceOf(msg.sender) > 0, "Insufficient amount for deposit");

File: contracts/Controller.sol [hunk 5]
```diff
@@ -1,22 +1,114 @@
...
+        require(approvedStrategies[_token][_strategy] == true, "!approved");
+
+        address _current = strategies[_token];
+        if (_current != address(0)) {
+            IStrategy(_current).withdrawAll();
+        }
+        strategies[_token] = _strategy;
+    }
+
+    function earn(address _token, uint256 _amount) public  override {
```
Comment (typicalbuster) [inline contracts/Controller.sol:80]: why 2 spaces between public and override? Same below

File: contracts/Controller.sol [hunk 6]
```diff
@@ -1,22 +1,114 @@
...
+        address _strategy = strategies[_token];
+       
+        IERC20(_token).safeTransfer(_strategy, _amount);
+        IStrategy(_strategy).deposit();
+    }
+
+    function balanceOf(address _token) external view  override returns (uint256) {
+        uint256 bal = 0;
+        if(strategies[_token] != address(0) ) {
```
Comment (typicalbuster) [inline contracts/Controller.sol:89]: strange codestyle using if(<condition> ), better if (<condition>). You can install https://github.com/prettier-solidity/prettier-plugin-solidity for solidity lint fixing.

File: contracts/ZunamiVault.sol [hunk 7]
```diff
@@ -1,13 +1,101 @@
...
+        _mint(msg.sender, shares);
+    }
+
+    function withdrawAll() external {
+        withdraw(balanceOf(msg.sender));
+    }
+
+    // Allow user to withdraw their initial deposit and burn the LP Tokens
+    function withdraw(uint256 _shares) public {
+        uint256 r = (balance().mul(_shares)).div(totalSupply());
```
Comment (typicalbuster) [inline contracts/ZunamiVault.sol:79]: Don't use single letter variable names, same everywhere, at least use shorts like "bal".

File: contracts/ZunamiVault.sol (hunk 4 above)
Comment (typicalbuster) [reply contracts/ZunamiVault.sol:56]: Yep

File: contracts/ZunamiVault.sol (hunk 3 above)
Comment (typicalbuster) [reply contracts/ZunamiVault.sol:73]: Yep

**Pull request closed without merging**